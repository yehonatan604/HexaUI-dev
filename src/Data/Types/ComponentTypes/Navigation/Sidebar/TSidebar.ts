import { IconType } from "react-icons";
import { TComponent, ComponentOptions } from "../../TComponent";

export type TSidebar = TComponent<"div"> & {
    options?: SidebarOptions;
    chevron?: IconType;
}

export type SidebarOptions = ComponentOptions & {
    maxWidth?: string;
    disableClose?: boolean;
}