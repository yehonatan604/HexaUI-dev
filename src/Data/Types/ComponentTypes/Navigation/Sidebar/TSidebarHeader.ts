import { TComponent, ComponentOptions } from "../../TComponent";

export type TSidebarHeader = TComponent<"div"> & {
    options?: SidebarHeaderOptions;
}

export type SidebarHeaderOptions = ComponentOptions & {
    placement: string;
}