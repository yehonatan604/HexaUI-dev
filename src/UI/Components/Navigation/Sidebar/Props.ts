import { TComponent } from "../../../..//Data/Types/TComponent";

export type SideMenuProps = TComponent<"div"> & {
    maxWidth?: string;
};