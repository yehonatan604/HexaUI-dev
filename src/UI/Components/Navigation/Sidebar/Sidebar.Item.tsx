import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { TSidebarItem } from "./types/TSidebarItem";
import Flex from "../../Layout/Flex/Flex";

const SidebarItem = (props: TSidebarItem) => {
  const { children, label, icon, options, className, innerClassName, ...componentProps } =
    props;

  // Options
  const labelBgColor = options?.labelBgVariant || "standard-l";
  const labelTextColor = options?.labelVariant || "standard-d";
  const hoverBgVariant = options?.hoverBgVariant || "standard-l";
  const hoverTextVariant = options?.hoverTextVariant || "standard-l";
  const textColor = options?.textVariant || "primary";
  const hover = `hover:bg-${options?.hoverBgVariant && hoverBgVariant} hover:text-${
    options?.hoverTextVariant && hoverTextVariant
  }`;

  // JSX
  return (
    <Flex
      options={{
        justify: FlexTypes.SpaceBetween,
        align: FlexTypes.Center,
      }}
      {...componentProps}
      className={`w-full ${className}`}
    >
      <Flex
        options={{
          justify: FlexTypes.Start,
          align: FlexTypes.Center,
        }}
        {...componentProps}
        className={`w-full p-1 pl-3 gap-5 text-${textColor} ${
          !options?.disableOpacityOnHover && "hover:opacity-60"
        } ${hover} cursor-pointer ${innerClassName}`}
      >
        {icon && icon}
        {children}
        {label && (
          <Flex
            options={{
              justify: FlexTypes.Center,
              align: FlexTypes.Center,
            }}
            className={`p-2 ml-auto mr-7 h-[25px] bg-${labelBgColor} text-${labelTextColor} rounded-lg`}
          >
            {label}
          </Flex>
        )}
      </Flex>
    </Flex>
  );
};

export default SidebarItem;
