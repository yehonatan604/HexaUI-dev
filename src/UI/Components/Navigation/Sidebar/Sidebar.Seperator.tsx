import { TSidebarSeperator } from "./types/TSidebarSeperator";
import Hr from "../../Typography/Hr/Hr";

const SidebarSeperator = (props: TSidebarSeperator) => {
  const { ...componentProps } = props;

  // JSX
  return (
    <Hr
      options={{
        border: {
          thickness: "5/6",
          variant: "standard",
        },
      }}
      {...componentProps}
    />
  );
};

export default SidebarSeperator;
