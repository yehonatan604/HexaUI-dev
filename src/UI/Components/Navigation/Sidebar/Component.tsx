import { useState } from "react";
import { FiChevronsRight } from "react-icons/fi";
import SidebarSeperator from "./Sidebar.Seperator/Component";
import SidebarItem from "./Sidebar.Item/Component";
import SidebarHeader from "./Sidebar.Header/Component";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import Flex from "../../Layout/Flex/Component";
import { TSidebar } from "../../../../Data/Types/ComponentTypes/Navigation/Sidebar/TSidebar";
import useTheme from "../../../../Core/Hooks/useTheme";

const Sidebar = (props: TSidebar) => {
  const { children, options, className, chevron: icon, ...componentProps } = props;

  // Hooks
  const [isOpen, setIsOpen] = useState(true);
  const { colors } = useTheme();

  // Options
  const chevronColor = options?.textVariant || colors.standradLight;
  const transition = "transition-all duration-1000";
  const maxWidth = options?.maxWidth || "w-[280px]";
  const bgColor = options?.bgVariant || colors.standradDark;
  const borderColor = options?.border?.variant || colors.standradDark;
  const borderThickness = options?.border?.thickness || 4;
  const border = `border-r-${borderThickness} border-${borderColor}`;

  // Functions
  const toggleMenu = () => setIsOpen(!isOpen);

  // JSX
  const Chevron = icon ? icon : FiChevronsRight;

  return (
    <Flex>
      <Flex
        options={{
          direction: FlexDir.Col,
        }}
        {...componentProps}
        className={`${isOpen ? `${maxWidth} max-md:w-[95vw]` : "w-[0px]"} bg-${bgColor} ${
          !isOpen && "opacity-5"
        } ${transition} duration-1000 overflow-hidden text-nowrap text-ellipsis gap-4 pb-4 ${
          options?.border && border
        } ${className}`}
      >
        {children}
      </Flex>
      {!options?.disableClose && (
        <Chevron
          onClick={toggleMenu}
          className={`text-2xl ${chevronColor} cursor-pointer ${
            isOpen && "transform rotate-180"
          } ${transition} `}
        />
      )}
    </Flex>
  );
};

Sidebar.Seperator = SidebarSeperator;
Sidebar.Item = SidebarItem;
Sidebar.Header = SidebarHeader;

export default Sidebar;
