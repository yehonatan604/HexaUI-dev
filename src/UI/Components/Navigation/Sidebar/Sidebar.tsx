import { useState } from "react";
import { FiChevronsRight } from "react-icons/fi";
import SidebarSeperator from "./Sidebar.Seperator";
import SidebarItem from "./Sidebar.Item";
import SidebarHeader from "./Sidebar.Header";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import Flex from "../../Layout/Flex/Flex";
import { TSidebar } from "./types/TSidebar";
import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";

const Sidebar = (props: TSidebar) => {
  const { children, options, className, chevron: icon, ...componentProps } = props;

  // Hooks
  const [isOpen, setIsOpen] = useState(true);
  const { colors } = useTheme();

  // Options
  const chevronColor = options?.textVariant || colors.standradLight;
  const transition = "transition-all duration-1000";
  const maxWidth = options?.maxWidth || "w-[280px]";
  const bgColor = options?.bgVariant || colors.standradDark;
  const borderColor = options?.border?.variant || colors.standradDark;
  const borderThickness = options?.border?.thickness || 4;
  const border = `border-r-${borderThickness} border-${borderColor}`;

  // Functions
  const toggleMenu = () => setIsOpen(!isOpen);

  // JSX
  const Chevron = icon ? icon : FiChevronsRight;

  return (
    <Flex>
      <Flex
        options={{
          direction: FlexDir.Col,
        }}
        {...componentProps}
        className={`${isOpen ? `${maxWidth} max-md:w-[95vw]` : "w-[0px]"} bg-${bgColor} ${
          !isOpen && "opacity-5"
        } ${transition} duration-1000 overflow-hidden text-nowrap text-ellipsis gap-4 pb-4 ${
          options?.border && border
        } ${className}`}
      >
        {children}
      </Flex>
      {!options?.disableClose && (
        <Chevron
          onClick={toggleMenu}
          className={`text-2xl ${chevronColor} cursor-pointer ${
            isOpen && "transform rotate-180"
          } ${transition} `}
        />
      )}
    </Flex>
  );
};

Sidebar.Seperator = SidebarSeperator;
Sidebar.Item = SidebarItem;
Sidebar.Header = SidebarHeader;

export default Sidebar;
