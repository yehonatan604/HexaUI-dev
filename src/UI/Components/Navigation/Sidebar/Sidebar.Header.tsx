import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { TSidebarHeader } from "./types/TSidebarHeader";
import Flex from "../../Layout/Flex/Flex";

const SidebarHeader = (props: TSidebarHeader) => {
  const { children, options, ...componentProps } = props;

  // Options
  const bgColor = options?.bgVariant || "standard";
  const textColor = options?.textVariant || "standard-l";

  // JSX
  return (
    <Flex
      options={{
        justify: options?.placement,
        align: FlexTypes.Center,
      }}
      {...componentProps}
      className={`p-4 w-full bg-${bgColor} text-${textColor}`}
    >
      {children}
    </Flex>
  );
};

export default SidebarHeader;
