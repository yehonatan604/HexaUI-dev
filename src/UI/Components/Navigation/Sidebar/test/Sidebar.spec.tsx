import { render, screen, fireEvent } from "@testing-library/react";
import Sidebar from "../Sidebar"; // Adjust the import path
import { describe, it, expect } from "vitest";
import { FiChevronsRight } from "react-icons/fi";

describe("Sidebar Component", () => {
  it("should render children correctly", () => {
    render(
      <Sidebar options={{}}>
        <div>Sidebar Child 1</div>
        <div>Sidebar Child 2</div>
      </Sidebar>
    );

    // Check if both children are rendered
    expect(screen.getByText("Sidebar Child 1")).toBeInTheDocument();
    expect(screen.getByText("Sidebar Child 2")).toBeInTheDocument();
  });

  it("should apply custom background color and max width", () => {
    render(
      <Sidebar options={{ bgVariant: "blue-500", maxWidth: "w-80" }}>
        <div>Sidebar Content</div>
      </Sidebar>
    );

    const sidebarElement = screen.getByText("Sidebar Content").parentElement;

    // Check if the custom background and width are applied
    expect(sidebarElement).toHaveClass("bg-blue-500", "w-80");
  });

  it("should toggle sidebar visibility when the chevron is clicked", () => {
    const { container } = render(
      <Sidebar options={{}} chevron={FiChevronsRight}>
        <div>Sidebar Content</div>
      </Sidebar>
    );

    const chevronIcon = container.querySelector("svg");
    const sidebarElement = screen.getByText("Sidebar Content").parentElement;

    // Check if sidebar is open initially
    expect(sidebarElement).toHaveClass("w-[280px]");

    // Simulate clicking the chevron icon
    fireEvent.click(chevronIcon!);

    // Check if sidebar closes
    expect(sidebarElement).toHaveClass("w-[0px]");
  });

  it("should disable sidebar closing when disableClose is set", () => {
    render(
      <Sidebar options={{ disableClose: true }}>
        <div>Sidebar Content</div>
      </Sidebar>
    );

    const sidebarElement = screen.getByText("Sidebar Content").parentElement;
    const chevronIcon = screen.queryByRole("button");

    // Check that the chevron icon is not rendered when disableClose is true
    expect(chevronIcon).toBeNull();

    // Check if the sidebar remains open
    expect(sidebarElement).toHaveClass("w-[280px]");
  });

  it("should render Sidebar.Item with icon and label", () => {
    render(
      <Sidebar options={{}}>
        <Sidebar.Item label="Item Label" icon={<FiChevronsRight />}>
          Sidebar Item
        </Sidebar.Item>
      </Sidebar>
    );

    // Check if the item and its label are rendered
    expect(screen.getByText("Sidebar Item")).toBeInTheDocument();
    expect(screen.getByText("Item Label")).toBeInTheDocument();

    // Check if the icon is rendered
    expect(screen.getByTestId("chevron-icon")).toBeInTheDocument();
  });

  it("should render Sidebar.Header with custom background and text color", () => {
    render(
      <Sidebar options={{}}>
        <Sidebar.Header
          options={{ placement: "top", bgVariant: "failure", textVariant: "white" }}
        >
          Sidebar Header
        </Sidebar.Header>
      </Sidebar>
    );

    const headerElement = screen.getByText("Sidebar Header").parentElement;

    // Check if the custom background and text color are applied
    expect(headerElement.firstChild).toHaveClass("bg-failure");
  });

  it("should render and match the snapshot", () => {
    const { asFragment } = render(
      <Sidebar
        options={{
          bgVariant: "blue-500",
          maxWidth: "w-80",
        }}
      >
        <Sidebar.Header
          options={{ placement: "top", bgVariant: "gray-800", textVariant: "white" }}
        >
          Header
        </Sidebar.Header>
        <Sidebar.Item label="Item Label" icon={<FiChevronsRight />}>
          Item Content
        </Sidebar.Item>
        <Sidebar.Seperator />
      </Sidebar>
    );

    // Snapshot the Sidebar structure
    expect(asFragment()).toMatchSnapshot();
  });
});
