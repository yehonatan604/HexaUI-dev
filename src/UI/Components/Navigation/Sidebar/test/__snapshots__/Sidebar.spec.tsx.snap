// Vitest Snapshot v1

exports[`Sidebar Component > should render and match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row bg- text- justify-start items-start flex-nowrap undefined"
  >
    <div
      class="flex flex-col bg- text- justify-start items-start flex-nowrap w-80 max-md:w-[95vw] bg-blue-500 false transition-all duration-1000 duration-1000 overflow-hidden text-nowrap text-ellipsis gap-4 pb-4 undefined undefined"
    >
      <div
        class="flex flex-row bg- text- justify-top items-center flex-nowrap p-4 w-full bg-gray-800 text-white"
      >
        Header
      </div>
      <div
        class="flex flex-row bg- text- justify-between items-center flex-nowrap w-full undefined"
      >
        <div
          class="flex flex-row bg- text- justify-start items-center flex-nowrap w-full p-1 pl-3 gap-5 text-primary hover:opacity-60 hover:bg-undefined hover:text-undefined cursor-pointer undefined"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="13 17 18 12 13 7"
            />
            <polyline
              points="6 17 11 12 6 7"
            />
          </svg>
          Item Content
          <div
            class="flex flex-row bg- text- justify-center items-center flex-nowrap p-2 ml-auto mr-7 h-[25px] bg-standard-l text-standard-d rounded-lg"
          >
            Item Label
          </div>
        </div>
      </div>
      <hr
        class="border-1 border-standard w-1/2 m-auto my-2 undefined"
      />
    </div>
    <svg
      class="text-2xl standard-l cursor-pointer transform rotate-180 transition-all duration-1000 "
      data-testid="chevron-icon"
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="13 17 18 12 13 7"
      />
      <polyline
        points="6 17 11 12 6 7"
      />
    </svg>
  </div>
</DocumentFragment>
`;
